File: java-mvc/hexagonal-architecture/src/main/java/com/ecommerce/hexagonal/adapter/input/web/dto/CreateProductRequest.java
package com.ecommerce.hexagonal.adapter.input.web.dto;

import jakarta.validation.constraints.*;
import java.math.BigDecimal;

public record CreateProductRequest(
        @NotBlank(message = "Name is required")
        @Size(max = 255, message = "Name cannot exceed 255 characters")
        String name,

        String description,

        @NotNull(message = "Price is required")
        @DecimalMin(value = "0.01", message = "Price must be greater than zero")
        BigDecimal price,

        @NotBlank(message = "Category is required")
        String category,

        @NotNull(message = "Stock quantity is required")
        @Min(value = 0, message = "Stock cannot be negative")
        Integer stockQuantity,

        String sku
) {}

File: dotnet-mvc/HexagonalArchitecture/Adapters/Input/Web/Dto/CreateProductRequest.cs
using System.ComponentModel.DataAnnotations;

namespace HexagonalArchitecture.Adapters.Input.Web.Dto;

/// <summary>
/// DTO for creating a new product. Belongs to the INPUT ADAPTER layer.
/// Uses DataAnnotations for basic request validation.
/// </summary>
public record CreateProductRequest(
    [Required] string Name,
    string? Description,
    [Required][Range(0.01, double.MaxValue, ErrorMessage = "Price must be greater than zero.")] decimal Price,
    [Required] string Category,
    [Required][Range(0, int.MaxValue)] int StockQuantity,
    [Required] string Sku
);

/// <summary>
/// DTO for updating an existing product. Belongs to the INPUT ADAPTER layer.
/// </summary>
public record UpdateProductRequest(
    [Required] string Name,
    string? Description,
    [Required][Range(0.01, double.MaxValue, ErrorMessage = "Price must be greater than zero.")] decimal Price,
    [Required] string Category,
    [Required] string Sku
);
